Jammyjellyfish2204 nextcloud24

stvhay opened the pull request:
This is my best attempt to get MIAB up to NC 24 for the Jammy release.

stvhay commented:
- Nginx is configured to respond to webfinger and nodeinfo requests.
- config.php has to be adjusted during the update to allow the update to complete. This is due to maybe a bug fix on the strictness of the read_only config setting.
- an imagemagick library is added to get NC to stop complaining about svg support.
- logic is added to set a telephone locale, which NC needs to guess country codes.
- fixed overwrite.cli.url. which NC now wants the full URL.


Notes: There are quite a few apps that could be disabled in NC. Not sure how much it is worth it to pick through them.

stvhay commented:
So it turns out occ provides a way to change config parameters and I think I should patch the implementation to use it before we merge. Much cleaner.

yodax commented on setup/nextcloud.sh:
```diff
@@ -328,11 +289,57 @@ hide_output sudo -u www-data php$PHP_VER /usr/local/lib/owncloud/console.php app
+TIMEZONE=$(cat /etc/timezone)
+instanceid=oc$(echo $PRIMARY_HOSTNAME | sha1sum | fold -w 10 | head -n 1)
+CONFIG_TEMP=/tmp/cfg-$instanceid.json
+PHONE_REGION=$(locale | grep TELEPHONE | sed -e 's/LC_.*=".*_//' | sed -e 's/\..*//')
```
For me this results in "C as the region. The " causes a problem on upgrading since it's invalid json. However C as a region isn't valid ISO 3166-1, which is the format NC requires: https://docs.nextcloud.com/server/stable/admin_manual/configuration_server/config_sample_php_parameters.html?highlight=default_phone_region

I've removed it on my instance and left the default value.

stvhay replied:
Without this, nextcloud was giving me a warning. For me, I am extracting 'US' from the locale.

I might need to find a more reliable way to get this code. I'll do some research on it.

yodax replied:
```locale
LANG=C.UTF-8
LANGUAGE=
LC_CTYPE="C.UTF-8"
LC_NUMERIC="C.UTF-8"
LC_TIME="C.UTF-8"
LC_COLLATE="C.UTF-8"
LC_MONETARY="C.UTF-8"
LC_MESSAGES="C.UTF-8"
LC_PAPER="C.UTF-8"
LC_NAME="C.UTF-8"
LC_ADDRESS="C.UTF-8"
LC_TELEPHONE="C.UTF-8"
LC_MEASUREMENT="C.UTF-8"
LC_IDENTIFICATION="C.UTF-8"
LC_ALL=```

stvhay replied:
I have added to the code some checks if the locale is not set to something that provides a country code. In this case, it leaves the country code blank. This will cause a warning with nextcloud admin panel, but I don't think its too serious.

yodax commented on setup/nextcloud.sh:
```diff
@@ -206,39 +214,38 @@ if [ ! -f $STORAGE_ROOT/owncloud/owncloud.db ]; then
+			"overwrite.cli.url": "https://${PRIMARY_HOSTNAME}/cloud",
+			"user_backends": [
+				{
+					"class": "\\\OCA\\\UserExternal\\\IMAP",
```
Should be `"memcache.local": "\\\\OC\\\\Memcache\\\\APCu",`

stvhay replied:
So, the three backslashes works for me and validates, but I agree that 4 backslashes is clearer escaping so I'll make the change.

yodax replied:
Pretty strange that that validates, could be the locale too I guess. When I validate 3 backslashes in an online validator it fails, just like on the machine it self. Does 4 work for you too?

yodax commented on setup/nextcloud.sh:
```diff
@@ -206,39 +214,38 @@ if [ ! -f $STORAGE_ROOT/owncloud/owncloud.db ]; then
+					"arguments": [ "127.0.0.1", 143, null, null, false, false ]
+				}
+			],
+		  "memcache.local": "\\\OC\\\Memcache\\\APCu",
```
Should be `"class": "\\\\OCA\\\\UserExternal\\\\IMAP",`

stvhay replied:
Again worked for me, but I will make the change because I think its clearer escaping.

yodax commented:
Great work on this!

We're going to need this in there (iI've run it manually):

`sudo -u www-data php$PHP_VER /usr/local/lib/owncloud/occ db:add-missing-primary-keys`

This is the diff for the config update: 

```diff --git a/setup/nextcloud.sh b/setup/nextcloud.sh
index df6be86..f22467a 100755
--- a/setup/nextcloud.sh
+++ b/setup/nextcloud.sh
@@ -322,18 +322,16 @@ cat > $CONFIG_TEMP <<-EOF
        "system": {
                "config_is_read_only": true,
                "trusted_domains": ["$PRIMARY_HOSTNAME"],
-         "memcache.local": "\\\OC\\\Memcache\\\APCu",
-         "mail_from_address": "administrator",
-         "logtimezone": "$TIMEZONE",
-         "logdateformat": "Y-m-d H:i:s",
-         "mail_domain": "$PRIMARY_HOSTNAME",
-         "default_phone_region": "$PHONE_REGION",
-               "user_backends": [
-                       {
-                               "class": "\\\OCA\\\UserExternal\\\IMAP",
-                               "arguments": [ "127.0.0.1", 143, null, null, false, false ]
-                       }
-               ]
+               "memcache.local": "\\\\OC\\\\Memcache\\\\APCu",
+               "overwrite.cli.url": "https://${PRIMARY_HOSTNAME}/cloud",
+               "mail_from_address": "administrator",
+               "logtimezone": "$TIMEZONE",
+               "logdateformat": "Y-m-d H:i:s",
+               "mail_domain": "$PRIMARY_HOSTNAME",
+               "user_backends": [{
+                       "class": "\\\\OCA\\\\UserExternal\\\\IMAP",
+                       "arguments": ["127.0.0.1", 143, null, null, false, false]
+               }]
        }
 }
 EOF```

Please check the initial config setup too. Those memcache and class values are not correct either. I don't think they will work during a clean install. I haven't tested this. To make it easier to debug you can grap the generated temp file and paste it in a web validator. 

When all of this is done the only thing remaining on my upgraded install is the SQLite warning and that's fine.

stvhay commented on setup/nextcloud.sh:
```diff
@@ -191,6 +191,14 @@ if [ ! -d /usr/local/lib/owncloud/ ] || [[ ! ${CURRENT_NEXTCLOUD_VER} =~ ^$nextc
 		fi
+
+		# Remove the read-onlyness of the config.
+		sed -ie '/config_is_read_only/d' $STORAGE_ROOT/owncloud/config.php
```
So the read-only config is something that was originally set and so I am trying to not change what was already there. That said, I think this option got more strict in a recent release and it means I need to undo the config_is_read_only parameter by directly manipulating the file.

It gets set back to true in the JSON upload of the final config.

I would be happy to set it back to false and keep it that way if @JoshData is ok with it.

stvhay commented:
@yodax  Thank you for taking the time to review and provide comments to my PR. I have pushed a revision to the code which I believe addresses all of your concerns. Let me know if you see anything else. 

Oh. One thing---I've run this from a fresh Ubuntu install and it worked fine. But one thing I think we could do once we get it merged is test the different upgrade paths someone might take or what we are going to support. Right now I've started from a fresh 22.04 install, so I've not looked at dist-upgrading to 22.04 and going from there.

yodax commented:
Not sure if you missed it, this needs to be added to the config update too:

`+               "overwrite.cli.url": "https://${PRIMARY_HOSTNAME}/cloud",`

Now it's only in the initial config setup.

yodax commented:
> dist-upgrading to 22.04

This is an unsupported upgrade path. So no worries about that one. I've tested a fresh install but with a restored backup of /home/user-data. 

Nice work!

stvhay commented:
> + "overwrite.cli.url": "https://${PRIMARY_HOSTNAME}/cloud",

Pushed.

yodax commented:
Awesome. I'll pull tomorrow and double check. But this looks good to me, I'll report back how it went!

stvhay commented:
> Awesome. I'll pull tomorrow and double check. But this looks good to me, I'll report back how it went!

Thanks!  If you are interested, I have a branch on my repo with all of my pending PRs and a roundcube update from @kiekerjan. This includes #2153 #2156 #2157 #2158 and #2159. I've been running it for a while and it seems good. But I understand if you want to be conservative since you are running this on your actual production machine. 

Anyways, here is the branch: https://github.com/stvhay/mailinabox/tree/jammyjellyfish2204-everything

yodax commented:
This worked for me. The telephone field is empty, which is fine.

I just noticed an extra error in the log:

`Error: stepping, UNIQUE constraint failed: oc_users_external.uid, oc_users_external.backend (19)`

```
sqlite> select * from oc_users_external;
michael@kroes.email|{localhost:993/imap/ssl/novalidate-cert}|Michael Kroes
michael@kroes.email|127.0.0.1|
*** other users ommited **
```

I think it's due to this line in the original Jammy PR by @JoshData : https://github.com/mail-in-a-box/mailinabox/pull/2083/files#diff-82e1be7bfd6ea28b0bc7b8ee6ae4a78dcaea281159e0853b3039b97fc231cbceR360

I trashed the 127.* records and updated the other records to have 127.0.0.1 as the backend. Then it runs without errors. 

Could you check your table and see if it migrated correctly? If this is a problem we can solve it separately, it has nothing to do with the PR itself.

stvhay commented:
> Could you check your table and see if it migrated correctly? If this is a problem we can solve it separately, it has nothing to do with the PR itself.

```
sqlite> select * from oc_users_external ;
127.0.0.1|me@leptonmail.com|
sqlite> 
```

Mine looks fine, but it is probably because I'm coming from a fresh install.

yodax commented:
Thanks for checking. For me this PR is good 👍 

@JoshData did you see any duplicates in that table when you upgraded your instance? I'm happy to blame this on something that happened on my machine alone. If more users report it, we could write a script that fixes it.

JoshData commented:
Yeah I had duplicates.

stvhay commented:
I have a patch developed to update all the wget_verify downloads to SHA256 hashes. We want to NSA-proof our email after all :). Should I include it here or wait until its merged and do a separate PR? The scope includes zpush, roundcube, and management as well.

kiekerjan commented on setup/nextcloud.sh:
```diff
@@ -206,39 +215,38 @@ if [ ! -f $STORAGE_ROOT/owncloud/owncloud.db ]; then
+	}
 EOF
 
+	sudo -u www-data php8.0 /usr/local/lib/owncloud/occ config:import $CONFIG_TEMP
```
Use php$PHP_VER instead of php8.0? Also on line 355

stvhay replied:
thank you. fixed.

JoshData commented:
> I have a patch developed to update all the wget_verify downloads to SHA256 hashes. We want to NSA-proof our email after all :). Should I include it here or wait until its merged and do a separate PR?

Separate PR. Please don't make this harder for me. :|

stvhay commented:
Merged in jammyjellyfish2204 and cleaned merge conflicts. 

Potentially conflicts with #2162 due to:
- It contains updated SHA256 hashes, whereas this PR contains SHA-1 hashes.
- It holds SHA256 values for NC versions only up to 23.

stvhay commented:
I've tried to simplify this PR for ease of review and inclusion into the project:

- Taken out the changes to manipulate the config via the Nextcloud API.
- Removed the stripping down and disabling of additional NC functionality to just calendar and contacts plugins.

What I've kept:
- Added nginx rewrites to handle webfinger and nodeinfo requests to eliminate NC warnings
- Added libmagickcore-6.q16-6-extra package to eliminate NC SVG warnings
- Added db:add-missing-primary-keys to the db migration step
- The config parameter overwrite.cli.url is populated with the full URL, per NC warning
- The config parameter default_phone_region value is surmised from the locale, or left blank if the locale is not defined
- Added log rotataion for nextcloud.log. Rotates when log reaches 10M.

louwers commented:
Saw this during setup:

```
Installing Nextcloud (contacts/calendar)...─────────────────────Hostname────────────────────────────────────────┐
Nextcloud is already latest version needs a name, called a 'hostname'. The name will form a part of the box's   │  
Error: stepping, UNIQUE constraint failed: oc_users_external.uid, oc_users_external.backend
```

binarykitchen commented:
I am still seeing this error

```
Error: stepping, UNIQUE constraint failed: oc_users_external.uid, oc_users_external.backend (19)
```

What to do? Ran latest script on latest Ubuntu.

binarykitchen commented:
@yodax yes I would write such a script since it's happening to other users including me.
